src/internal/connector/onedrive/url_cache.go

```diff
@@ -0,0 +1,275 @@
+package onedrive
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/alcionai/clues"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+
+	"github.com/alcionai/corso/src/internal/common/ptr"
+	"github.com/alcionai/corso/src/internal/connector/graph"
+	"github.com/alcionai/corso/src/pkg/fault"
+	"github.com/alcionai/corso/src/pkg/logger"
+)
+
+type itemProperties struct {
+	downloadURL string
+	isDeleted   bool
+}
+
+// urlCache caches download URLs for drive items
+type urlCache struct {
+	driveID string
+	// urlMap stores Item ID -> item property map
+	urlMap          map[string]itemProperties
+	lastRefreshTime time.Time
+	refreshInterval time.Duration
+	// cacheLock protects urlMap and lastRefreshTime
+	cacheLock sync.RWMutex
+	// refreshMutex serializes cache refresh attempts
+	refreshMutex    sync.Mutex
+	deltaQueryCount int
+
+	svc           graph.Servicer
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager
+
+	errors *fault.Bus
+}
+
+// newURLache creates a new URL cache for the specified drive ID
+func newURLCache(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) (*urlCache, error) {
+	err := validateCacheParams(
+		driveID,
+		refreshInterval,
+		svc,
+		itemPagerFunc)
+	if err != nil {
+		return nil, clues.Wrap(err, "invalid cache parameters")
+	}
+
+	return &urlCache{
+			urlMap:          make(map[string]itemProperties),
+			lastRefreshTime: time.Time{},
+			driveID:         driveID,
+			refreshInterval: refreshInterval,
+			svc:             svc,
+			itemPagerFunc:   itemPagerFunc,
+			errors:          fault.New(false),
+		},
+		nil
+}
+
+// validateCacheParams validates the parameters passed to newURLCache
+func validateCacheParams(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) error {
+	if len(driveID) == 0 {
+		return clues.New("drive id is empty")
+	}
+
+	if refreshInterval < 0 {
```

Comment:
is 0 a valid refresh interval?  If so, what is the refresh rate?

Would it make sense for this to be some well defined minimum like, say, 5 minutes?  0 and 1 are relatively indifferent in this case.

Reply:
> is 0 a valid refresh interval? If so, what is the refresh rate?

Oops. This is a bug. Thanks for pointing out. 

> Would it make sense for this to be some well defined minimum like, say, 5 minutes? 

I want to allow specifying interval in seconds. This would allow us to test for stale cache conditions faster.